src/Discord.Net.Commands/Extensions/CommandServiceExtensions.cs

```diff
@@ -0,0 +1,28 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Discord.Commands
+{
+    public static class CommandServiceExtensions
+    {
+        public static async Task<IReadOnlyCollection<CommandInfo>> GetExecutableCommandsAsync(this IEnumerable<CommandInfo> commands, ICommandContext context, IServiceProvider provider)
+        {
+            var executableCommands = new List<CommandInfo>();
+
+            foreach (var command in commands)
+            {
+                var result = await command.CheckPreconditionsAsync(context, provider).ConfigureAwait(false);
+                if (result.IsSuccess)
+                    executableCommands.Add(command);
+            }
+
+            return executableCommands;
+        }
+        public static Task<IReadOnlyCollection<CommandInfo>> GetExecutableCommandsAsync(this CommandService commandService, ICommandContext context, IServiceProvider provider)
+            => GetExecutableCommandsAsync(commandService.Commands, context, provider);
+        public static Task<IReadOnlyCollection<CommandInfo>> GetExecutableCommandAsync(this ModuleInfo module, ICommandContext context, IServiceProvider provider)
+            => GetExecutableCommandsAsync(module.Commands.Concat(module.Submodules.SelectMany(sm => sm.Commands)), context, provider);
```

Comment:
This only recurses through the first level of commands; it won't work for something like this:
```cs
class ManageCommands
{
    class ManageAdminCommands
    {
        class ManageInviteLimitCommands
        {
            // commands here
        }
    }
}
```

Reply:
Dang it, I had that thought in my head and forgot about it hours later when I actually opened VS again.